parser/Parser.cc

```diff
@@ -111,25 +111,22 @@ unique_ptr<Node> Parser::run(core::GlobalState &gs, core::FileRef file, Parser::
     // Marked `const` so that we can re-use across multiple `build()` invocations
     const Builder builder(gs, file);
 
-    // `buffer` and `scratch` can also be shared by both drivers
-    auto source = file.data(gs).source();
+    // `source` and `scratch` can also be shared by both drivers
+    const auto source = file.data(gs).source();
     // The lexer requires that its buffers end with a null terminator, which core::File
     // does not guarantee.  Parsing heredocs for some mysterious reason requires two.
-    string buffer;
-    buffer.reserve(source.size() + 2);
-    buffer += source;
-    buffer += "\0\0"sv;
+    ENFORCE(source.size() >= 2 && string_view(source.data() + source.size() - 2, 2) == "\0\0");
```

Comment:
This is the part that was previously duplicating the source buffer (allocating into `buffer`)